### PR TITLE
reworked the main loop for protobrain to be simpler

### PR DESCRIPTION
**Pull request by jedmccaleb**

### CODE REVIEW

**Comment by jedmccaleb** [comment]
ok this is ready now. Probably some dead code to clean up from functions no longer used

**Review by wumpusman — APPROVED** [review]
I'm assuming protobrain and fworld are still connecting.

File: models/integrated/protobrain/protobrain.go [hunk 1]
```diff
@@ -143,6 +137,40 @@ func (ss *Sim) NewRun() {
...
-func startWorkerLoop(sim *Sim) {
+func (sim *Sim) OnObserve() {
+	for _, name := range sim.Net.LayersByClass(emer.Input.String()) {
+		agent.AgentApplyInputs(sim.Net.AsAxon(), sim.WorldEnv, name)
+	}
+}
+
+func (sim *Sim) OnStep(obs map[string]etensor.Tensor) map[string]agent.Action {
+	sim.NumSteps++
+	if sim.NumSteps >= gConfig.LIFETIME {
```
Comment (qemqemqem) [inline models/integrated/protobrain/protobrain.go:148]: Is this going to be adaptable to the case where the simulation ends early?

File: models/integrated/protobrain/protobrain.go [hunk 2]
```diff
@@ -143,6 +137,40 @@ func (ss *Sim) NewRun() {
...
+	// start the server listening for the world telling you to step
+	// 		Step Called by the world:
+	// 		increment the count of steps
+	// 		run the lopper
+	// 		get the action from the current brain state
+	// 		return the action to the world
+	// write results after LIFETIME steps
+	// exit
+
+	agent.StartServer("127.0.0.1:9090", sim.OnStep)
```
Comment (qemqemqem) [inline models/integrated/protobrain/protobrain.go:174]: This should come from the config

File: models/integrated/protobrain/protobrain.go [hunk 3]
```diff
@@ -143,6 +137,40 @@ func (ss *Sim) NewRun() {
...
+	sim.WorldEnv.SetObservations(obs)
+
+	log.Info("OnStep: ", sim.NumSteps)
+	sim.Loops.Step(sim.Loops.Mode, 1, etime.Trial)
+	actions := agent.GetAction(sim.Net.AsAxon())
+
+	return actions
+}
+
+func (sim *Sim) startWorkerLoop() {
```
Comment (qemqemqem) [inline models/integrated/protobrain/protobrain.go:164]: This function should go in a library

File: models/integrated/protobrain/protobrain.go [hunk 4]
```diff
@@ -143,6 +137,40 @@ func (ss *Sim) NewRun() {
...
 }
 
-func startWorkerLoop(sim *Sim) {
+func (sim *Sim) OnObserve() {
+	for _, name := range sim.Net.LayersByClass(emer.Input.String()) {
+		agent.AgentApplyInputs(sim.Net.AsAxon(), sim.WorldEnv, name)
+	}
+}
+
+func (sim *Sim) OnStep(obs map[string]etensor.Tensor) map[string]agent.Action {
```
Comment (qemqemqem) [inline models/integrated/protobrain/protobrain.go:146]: Can this go in a library too?

File: models/integrated/protobrain/protobrain.go (hunk 1 above)
Comment (jedmccaleb) [reply models/integrated/protobrain/protobrain.go:148]: What is causing it to end early?

File: models/integrated/protobrain/protobrain.go (hunk 3 above)
Comment (jedmccaleb) [reply models/integrated/protobrain/protobrain.go:164]: hmm maybe if we end up having lots of projects that need it

File: models/integrated/protobrain/protobrain.go (hunk 4 above)
Comment (jedmccaleb) [reply models/integrated/protobrain/protobrain.go:146]: hmm maybe if we end up having lots of projects that need it

File: models/integrated/protobrain/protobrain.go (hunk 1 above)
Comment (qemqemqem) [reply models/integrated/protobrain/protobrain.go:148]: Like if the agent dies or something.

**Pull request merged**